File: src/AddressBook/AddressBook.java
package AddressBook;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Collections;

public class AddressBook  implements Cloneable
{
    //디폴트생성자
    public AddressBook()
    {
        this.personals = new ArrayList<Personal>();
    }
    //복사생성자
    public AddressBook(AddressBook source)
    {
        //깊은 복사
        Personal deepCopyPersonal = null;
        //기존의 원본을 clone하여 ArrayList를 생성한 뒤에 새로운 객체로 배열요소 수정하기
        this.personals = (ArrayList<Personal>)source.personals.clone();
        for(int i = 0; i < this.personals.size(); i++)
        {
            deepCopyPersonal = new Personal(source.getAt(i));
            this.personals.set(i, deepCopyPersonal);
        }
        /*
        //새로 ArrayList 객체를 생성한 다음 새로은 객체를 배열에 추가하기
        this.personals = new ArrayList<Personal>();
        for (Personal personal : source.personals)
        {
               deepCopyPersonal = new Personal(personal);
               this.personals.add(deepCopyPersonal);
        }
         */
    }
    //personals 가져오기
    public ArrayList<Personal> getPersonals()
    {
        return this.personals;
    }
    //Personal 객체 정보 기재하기
    public  int record(String name, String address, String telephoneNumber, String emailAddress)
    {
        int index = -1;//기재하기 실패할 경우 -1반환
        Personal personal = new Personal(name, address, telephoneNumber, emailAddress);
        boolean doesRecordingSucceed = this.personals.add(personal);
        if(doesRecordingSucceed == true)
        {
            index = this.personals.indexOf(personal);
        }
        return index;
    }
    //Personal 객체 정보 가져오기
    public Personal getAt(int index)
    {
        return this.personals.get(index);
    }
    /*
    //name 으로 Personal 객체 찾기
    public int find(String name, int[] indexes)
    {
        //ArrayList 중에서 같은 이름의 배열요소의 위치를 담을 공간
        indexes = new int[this.personals.size()];
        //ArrayList 의 처음부터 끝까지 반복을 돌리면서 찾으려는 이름이 있는지 확인하기
        int index = 0;//위치배열첨자
        int count = 0;//위치배열에 저장되는 개수
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //위치배열 첨자를 증가시킨다.
                index++;
                //위치배열에 저장되는 개수를 증가시킨다.
                count++;
            }
        }
        return count;
    }
     */
    /*
    public int[] find(String name, Integer count)
    {
        //ArrayList 중에서 같은 이름의 배열요소의 위치를 담을 공간
        int[] indexes = new int[this.personals.size()];
        //ArrayList 의 처음부터 끝까지 반복을 돌리면서 찾으려는 이름이 있는지 확인하기
        count = 0;//같은 배열 개수
        int index = 0;//배열 첨자
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //위치배열 첨자를 증가시킨다.
                index++;
                //개수를 증가시킨다
                count++;
            }
        }
        return indexes;
     */
    //ArrayList는 배열과 달리 capacity가 유동적으로 바뀌기 때문에 ArrayList를 이용하면
    //똑같은 이름을 찾는 행위를 한번만 하면됨.
    public ArrayList<Integer> find(String name)
    {
        //name으로 찾은 Personal 객체들의 위치를 저장할 위치ArrayList 생성하기
        ArrayList<Integer> indexes = new ArrayList<>();
        //주소록의 ArrayList 의 처음부터 끝까지 반복을 돌리면서 name과 같은 배열요소의 위치를 저장한다.
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            //해당 위치의 Personal객체를 구한다.
            personal = this.personals.get(i);
            //해당 Personal의 name이 찾고자 하는 name과 같으면
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes.add(i);
            }
        }
        return indexes;
    }
    /*
    //int[]배열을 이용하기 대신 배열생성 전에 크기를 구하기 위해 찾고자 하는 이름과 같은 이름이 배열에 있는지
    //확인하여 그 개수를 새고 배열을 작성한 뒤에 다시 똑같은 이름을 찾으면서 그 위치를 배열요소에 저장
    //똑같은 이름을 찾는 행위 2번 중복되어 시간낭비가 심함.
    public int[] find(String name)
    {

        //시간 절약 메모리 낭비
        //같은 이름이 없다고 초기화
        boolean isMatched = false;
        //ArrayList 중에서 같은 이름의 배열요소의 위치를 담을 공간
        int[] indexes = new int[this.personals.size()];
        //ArrayList 의 처음부터 끝까지 반복을 돌리면서 찾으려는 이름이 있는지 확인하기
        int index = 0;//배열첨자
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //위치배열 첨자를 증가시킨다.
                index++;
                //찾은게 있다고 표시한다.
                isMatched = true;
            }
        }
        if(isMatched == false)
        {
            indexes = null;
        }
        return indexes;
 */
/*
        //메모리 절약 하지만 시간이 많이 소모됨
        //배열값 초기화
        int[] indexes = null;
        int count = 0;//찾은 개수 초기화
        //ArrayList 의 처음부터 끝까지 찾고자 하는 이름과 비교하여 찾은 개수를 센다.
        for (Personal personal: this.personals)
        {
            if(name.compareTo(personal.getName())==0)
            {
                //찾은 개수를 증가시킨다.
                count++;
            }
        }
        //찾은게 있으면
        if(count > 0)
        {
            //배열을 할당한다.
            indexes = new int[count];
        }
        //ArrayList 의 다시 처음부터 끝까지 반복을 돌리면서 같은 배열요소의 위치를 저장한다.
        int index = 0;//배열첨자
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //배열첨자를 증가시킨다.
                index++;
            }
        }
        return indexes;
    }
    */

    //Personal 객체 정보 수정하기(주소, 전화번호, 이메일주소 변경 가능, 이름은 변경 불가)
    public int correct(int index, String address, String telephoneNumber, String emailAddress)
    {
        //새로운 객체가 생성되기 때문에 이렇게 하면 안됨
        // Personal personal = new Personal(this.personals.get(index).getName(), address,
        //        telephoneNumber, emailAddress);
        Personal personal = this.personals.get(index);
        personal.setAddress(address);
        personal.setTelephoneNumber(telephoneNumber);
        personal.setEmailAddress(emailAddress);
        //this.personals.set(index, personal);
        index = this.personals.indexOf(personal);
        return  index;
    }

    /*
    public Personal correct(int index, String address, String telephoneNumber, String emailAddress)
    {
        //Personal personal = new Personal(this.personals.get(index).getName(),
        //                           address, telephoneNumber, emailAddress);
        Personal personal = this.personals.get(index);
        personal.setAddress(address);
        personal.setTelephoneNumber(telephoneNumber);
        personal.setEmailAddress(emailAddress);
        //Personal oldPersonal = this.personals.set(index, personal);
        //return  oldPersonal;
        return personal;
    }
     */

    //Erase
    public int erase(int index)
    {
        Personal personal = this.personals.remove(index);
        if(personal != null)
        {
            index = -1;
        }
        return index;
    }
    //Arrange
    public void arrange()
    {
        Collections.sort(this.personals, new Comparator<Personal>() {
            @Override
            public int compare(Personal one, Personal other)
            {
                return one.getName().compareTo(other.getName());
            }
        });
    }
    @Override
    public AddressBook clone() throws CloneNotSupportedException
    {
        //깊은 복사
        //AddressBook addressBook = (AddressBook)super.clone();
        AddressBook addressBook = new AddressBook();
        Personal deepCopyPersonal = null;
        //기존의 원본을 clone하여 ArrayList를 생성한 뒤에 새로운 객체로 배열요소 수정하기
        //addressBook.personals = (ArrayList<Personal>)this.personals.clone();
        addressBook.personals = new ArrayList<Personal>();
        for(int i = 0; i < this.personals.size(); i++)
        {
            deepCopyPersonal = new Personal(this.getAt(i));
            //addressBook.personals.set(i, deepCopyPersonal);
            addressBook.personals.add(deepCopyPersonal);
        }
        return addressBook;
    }
    //인스턴스 멤버
    private ArrayList<Personal> personals;
}

File: AddressBook/C-AddressBook/C-Addressbook20200313/addressbook20200313/AddressBook.c
/***************************************************************************************************************
 파일이름: AddressBook.c
 기능: 주소록에 성명, 주소, 전화번호, 이메일주소를 기재하고, 이름을 통해 기재된
       주소록에서 정보를 찾고, 기재된 정보를 변경하거나 삭제하고 다시 정렬한다.
 작성자: 박인재
 작성일자: 2020/03/10
****************************************************************************************************************/
#include<stdio.h>
#include<string.h>
#include "AddressBook.h"
#pragma warning(disable:4996)
#if 0
int main(int argc, char* argv[]) {
    AddressBook addressBook = { {{"", "", "", ""}, {"", "", "", ""}, {"", "", "", ""},
                                 {"", "", "", ""}, {"", "", "", ""}, {"", "", "", ""}}, MAX, 0 };
    Long index;
    Long indexes[MAX];
    Long count;

    //1. 홍길동, 서울중구, 021766710, Hong을 기재한다.
    index = Record(&addressBook, "홍길동", "서울 중구", "021766710", "Hong");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //2. 고길동, 서울성동구, 025979576, Go를 기재한다.
    index = Record(&addressBook, "고길동", "서울 성동구", "025979576", "Go");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //3. 홍길동, 부산연수구, 0513267920, Hong2를 기재한다.
    index = Record(&addressBook, "홍길동", "부산 연수구", "0513267920", "Hong2");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //4. 정길동, 서울종로구, 024366751, Jung을 기재한다.
    index= Record(&addressBook, "정길동", "서울 종로구", "024366751", "Jung");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //5. 홍길동을 찾는다.
    Find(&addressBook, "홍길동", indexes, &count);
    index = 0;
    while (index<count)
    {
        printf("%d %s %s %s %s\n", indexes[index], addressBook.personals[indexes[index]].name, 
            addressBook.personals[indexes[index]].address, addressBook.personals[indexes[index]].telephoneNumber,
            addressBook.personals[indexes[index]].emailAddress);
        index++;
    }
    //6. 세번째 홍길동의 주소와 전화번호, 이메일 주소를 고친다.
    index = Correct(&addressBook, 2, "서울 광진구", "027971723", "Hong2");
    printf("%d %s %s %s\n", index, addressBook.personals[index].address, 
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //7. 고길동을 지운다.
    index = Erase(&addressBook, 1);
    if (index==-1)
    {
        printf("지워졌습니다.\n");
    }
    //8. 최길동, 서울성동구, 023517134, choi를 기재한다.
    index = Record(&addressBook, "최길동", "서울 성동구", "023517134", "choi");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //9. 고보건, 서울송파구, 021579134, Gobo를 기재한다.
    index = Record(&addressBook, "고보건", "서울 송파구", "021579134", "Gobo");
    printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
        addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
    //10. 고보건을 지운다.
    index = Erase(&addressBook, 5);
    if (index == -1)
    {
        printf("지워졌습니다.\n");
    }
    //11. 정리한다.
    Arrange(&addressBook);
    index = 0;
    while (index<addressBook.length)
    {
        printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
            addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
        index++;
    }

    return 0;
}
#endif
/***************************************************************************************************************
 함수명칭: Record
 기능: 주소록에 성명, 주소, 전화번호, 이메일주소를 기재한다.
 입력: 주소록, 성명, 주소, 전화번호, 이메일주소
 출력: 위치
****************************************************************************************************************/

Long Record(AddressBook* addressBook, char(*name), char(*address), char(*telephoneNumber), char(*emailAddress)) {
    Long index = -1;
    Long i = 0;

    //1. 성명, 주소, 전화번호, 이메일주소를 입력한다.
    //2. 빈 줄을 찾는다.
    while (i < addressBook->capacity && strcmp(addressBook->personals[i].name, "") != 0) {
        i++;
    }
    //3. 빈줄에 적는다.
    if (i < addressBook->capacity) {
        index = i;
        strcpy(addressBook->personals[index].name, name);
        strcpy(addressBook->personals[index].address, address);
        strcpy(addressBook->personals[index].telephoneNumber, telephoneNumber);
        strcpy(addressBook->personals[index].emailAddress, emailAddress);
        addressBook->length++;
    }
    //4. 적힌 줄의 위치를 출력한다.
    return index;
    //5. 끝내다.
}

/***************************************************************************************************************
 함수명칭: Find
 기능: 성명으로 찾고자 하는 정보를 찾는다.
 입력: 주소록, 성명
 출력: 위치들, 개수
****************************************************************************************************************/
void Find(AddressBook* addressBook, char(*name), Long(*indexes), Long* count) {
    Long i = 0;
    Long j = 0;

    *count = 0;
    //1. 성명을 입력한다.
    //2. 사용량보다 작거나 같을 동안 반복한다.
    while (i<addressBook->length)
    {
        //2.1 기재된 이름과 찾고자 하는 이름이 같으면
        if (strcmp(addressBook->personals[i].name, name)==0)
        {
            //2.1.1 개수를 센다.
            (*count)++;
            //2.1.2 위치를 저장한다.
            indexes[j] = i;
            j++;
        }
        i++;
    }
    //3. 위치들과 개수를 출력한다.
    //4. 끝내다.
}

/***************************************************************************************************************
 함수명칭: Correct
 기능: 찾고자 하는 위치와 변경할 내용을 입력한다.
 입력: 주소록, 위치, 주소, 전화번호, 이메일주소
 출력: 위치
****************************************************************************************************************/
Long Correct(AddressBook* addressBook, Long index, char(*address), char(*telephoneNumber), char(*emailAddress)) {
    //1. 위치, 주소, 전화번호, 이메일주소를 입력한다.
    //2. 위치에 적는다.
    strcpy(addressBook->personals[index].address, address);
    strcpy(addressBook->personals[index].telephoneNumber, telephoneNumber);
    strcpy(addressBook->personals[index].emailAddress, emailAddress);
    //3. 위치를 출력한다.
    return index;
    //4. 끝내다.
}

/***************************************************************************************************************
 함수명칭: Erase
 기능: 지울 위치의 이름을 지운다.
 입력: 주소록, 위치
 출력: 위치
****************************************************************************************************************/
Long Erase(AddressBook* addressBook, Long index) {
    //1. 지울 위치를 입력한다.
    //2. 위치를 지운다.
    strcpy(addressBook->personals[index].name, "X");
    index = -1;
    //3. 위치를 출력한다.
    return index;
    //4. 끝내다.
}

/***************************************************************************************************************
 함수명칭: Arrange
 기능: 주소록을 정렬한다.
 입력: 주소록
 출력: 없음
****************************************************************************************************************/
void Arrange(AddressBook* addressBook) {
    Personal personal;
    Long i = 1;
    Long j;
    Long count=0;

    //1. 사용량보다 작거나 같은동안 반복한다.
    while (i<addressBook->length)
    {
        //1.1 내림차순으로 정렬한다.
        personal = addressBook->personals[i];
        j = i - 1;
        while (j>=0 && strcmp(addressBook->personals[j].name, personal.name)<0)
        {
            addressBook->personals[j + 1] = addressBook->personals[j];
            j--;
        }
        addressBook->personals[j + 1] = personal;
        i++;
    }
    //2. 개수를 센다.
    i = 0;
    while (i<addressBook->length && strcmp(addressBook->personals[i].name, "X")!=0)
    {
        count++;
        i++;
    }
    //3. 오름차순으로 정렬한다.
    i = 1;
    while (i<count)
    {
        personal = addressBook->personals[i];
        j = i - 1;
        while (j>=0 && strcmp(addressBook->personals[j].name, personal.name)>0)
        {
            addressBook->personals[j + 1] = addressBook->personals[j];
            j--;
        }
        addressBook->personals[j + 1] = personal;
        i++;
    }
    //4. 끝내다.
}